### ISSUE TITLE
Delete installed apps as part of deleting a database?

### ISSUE THREAD

**Issue by per-samuelsson** [Starcounter/Starcounter#2786]
When working with #2684, I found that what causes it, is that when deleting a database, applications installed in that database - i.e. "playlists" - is not deleted as part of the deletion.

We need to decide how a "delete database" operation relate to installed applications. 

If they are to be deleted along with the database, I need help from @andwah to understand the full structure in order to be able to delete the files.

If we decide that applications should be untouched, we need to adapt the delete to that. As it is now, its not possible to recreate a database with the same name (#2684), since the remaining apps occupy the name/space for the same database.

As a side note, I also suggest that we should probably add a redirection to where installed applications are stored, just as with database data files, transaction logs and temporary files. The configuration file keeps a reference to installed applications, but the actual artifacts should be kept in a specific hive (I suggest "apps"). I can discuss this with @andwah and take it on me to redesign if he doesn't see any problems doing so.

@warpech, @dan31, WDYT about the title question?

**Comment by dan31** [comment]
If we view database approximately as an isolated VM that runs apps, then up to my feeling apps should be removed together with the database. This behavior would be consistent. @warpech?

**Comment by per-samuelsson** [comment]
> If we view database approximately as an isolated VM that runs apps, then up to my feeling apps should be removed together with the database. This behavior would be consistent.

This is my take too. I think we need to go this way, despite it will mean more effort when "recreating" databases, like @warpech tried in the first place, because he will need to reinstall all apps on every such cycle I think.

**Comment by warpech** [comment]
I agree that by default deleting the database should delete the apps. Just like in VM, deleting the image deletes the apps. Or in SQL, deleting the database deletes the stored procedures.

Then again, for dev purposes, I would find it very useful to be able to clear the database without deleting it. Should I create a ticket for that?

**Comment by miyconst** [comment]
>I would find it very useful to be able to clear the database without deleting it.

+1

**Comment by dan31** [comment]
We discussed the point with @per-samuelsson and came to the same conclusion: there is no practical reason to delete database physically if there will be a simple mean to vanish it (like you have it SQLite, MongoDB, whatever). @k-rus, there is a way today to DROP INDEX https://github.com/Starcounter/Starcounter/wiki/Index-Declaration, but is there a way to DROP TABLE? I assume there is no DELETE FROM today.

For conveniency it shouldn't be difficult to add a button to Administrator, which will iterate metadata and will clean up (but not delete) all tables that it can find in the metadata. Or drop those tables. @andwah, WDYT?

**Comment by andwah** [comment]
@dan31 it will not be so difficult to add this "Clean Data" button,  if there is a way to actually do the job (DROP TABLE, DROP INDEX etc..) on the server side.

**Comment by k-rus** [comment]
> @k-rus, there is a way today to DROP INDEX https://github.com/Starcounter/Starcounter/wiki/Index-Declaration, but is there a way to DROP TABLE? I assume there is no DELETE FROM today.

Only `DROP INDEX` exists. `DELETE FROM` does not work with large data sets.
From development point of view deleting database is good, since it guarantees to have clean database. Note that deleting any database object such as table, column, index requires to modify meta-data.
